Mathlib/Data/Multiset/DershowitzManna.lean

```diff
@@ -0,0 +1,409 @@
+      · change M = M - {a0} + {a0}
+        have this0: M = X + {a0} - {a} := by
+          rw [← h0, Multiset.sub_singleton, Multiset.erase_cons_head]
+        have a0M: a0 ∈ M := by
+          rw [this0, Multiset.sub_singleton, Multiset.mem_erase_of_ne]
+          rw [Multiset.mem_add, Multiset.mem_singleton]
+          · apply Or.inr
+            rfl
+          · exact fun h ↦ hyp (Eq.symm h)
+        rw [add_comm]
+        simp_all [Multiset.singleton_add]
+      exact h2
+
+lemma acc_cons [DecidableEq α] [Preorder α] (a : α) (M0 : Multiset α)
+    (_ : ∀ b M , LT.lt b a → AccM M → AccM (b ::ₘ M))
+    (_ : AccM M0)
+    (_ : ∀ M, MultisetRedLT M M0 → AccM (a ::ₘ M)) :
+    AccM (a ::ₘ M0) := by
+  constructor
+  intros N N_lt
+  change AccM N
+  rcases (red_insert N_lt) with ⟨x, H, h0⟩
+  case h.intro.inr h =>
+    rcases h with ⟨H, h0⟩
+    rw [H]
+    clear H
+    induction x using Multiset.induction with
+    | empty =>
+      simpa
+    | cons h =>
+      simp_all only [Multiset.mem_cons, or_true, implies_true, true_implies, forall_eq_or_imp,
+        Multiset.add_cons]
+  case h.intro.inl.intro =>
+    simp_all
+
+lemma acc_cons_of_acc [DecidableEq α] [Preorder α] (a : α)
+    (H : ∀ (b : α), ∀ M, LT.lt b a → AccM M → AccM (b ::ₘ M)) :
+    ∀ M, AccM M → AccM (a ::ₘ M) := by
+  unfold AccM
+  intros M h0
+  induction h0 with
+  | intro x wfH wfh2 =>
+    apply acc_cons
+    · simpa
+    · constructor
+      simpa
+    · simpa
+
+lemma acc_cons_of_acc_of_lt [DecidableEq α] [Preorder α] :
+    ∀ (a:α), Acc LT.lt a → ∀ M, AccM M → AccM (a ::ₘ M) := by
+  intro w w_a
+  induction w_a with
+  | intro x _ ih =>
+      intro M accM1
+      apply @acc_cons_of_acc α _ _ _ _ _ accM1
+      simp_all
+
+/- If all elements of a multiset `M` are accessible with `LT.lt`, then the multiset M is
+accessible given the `MultisetRedLT` relation. -/
+lemma acc_of_acc_lt [DecidableEq α] [Preorder α] :
+    ∀ (M : Multiset α), (∀x, x ∈ M → Acc LT.lt x) → AccM M  := by
+  intros M wf_el
+  induction M using Multiset.induction_on with
+  | empty =>
+    constructor
+    intro y y_lt
+    absurd y_lt
+    apply not_redLT_zero
+  | cons _ _ ih =>
+    apply acc_cons_of_acc_of_lt
+    · apply wf_el
+      simp_all
+    · apply ih
+      intros
+      apply wf_el
+      simp_all
+
+/- If `LT.lt` is well-founded, then `MultisetRedLT` is well-founded. -/
+lemma redLT_wf [DecidableEq α] [Preorder α]
+    (wf_lt : WellFoundedLT α) : WellFounded (MultisetRedLT : Multiset α → Multiset α → Prop) := by
+  constructor
+  intros a
+  apply acc_of_acc_lt
+  intros x _
+  apply wf_lt.induction x
+  intros y h
+  apply Acc.intro y
+  assumption
+
+/- If `MultisetRedLT` is well-founded, then its transitive closure `MultisetTransLT` is also
+well-founded. -/
+lemma transLT_wf [DecidableEq α] [LT α]
+    (h : WellFounded (MultisetRedLT : Multiset α → Multiset α → Prop)) :
+    WellFounded (MultisetTransLT : Multiset α → Multiset α → Prop) := by
+  unfold MultisetTransLT
```

Comment:
is this necessary? (would the `apply` below not work without it)?

Reply:
You are right. It is not necessary;-)